### PR TITLE
Point email action improvements

### LINKED ISSUE #7320: problem/feature with repeat point action
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
When you create a point action for opening an email, you can choose to repeat the action. That means if the contact opens different emails, he will be scored.
The thing is, he will also be scored if he opens multiple times the same email. So he will get more point than he really should have.

It is nice to have the possibility to repeat the action but it could be better if we could limit to add points for the 1st email.
| Q   | A
| --- | ---
| Mautic version | 2.15
| PHP version | 
| Browser | 

### Steps to reproduce
1. create a point action, choose opens an email. Enable to repeat the action.
2. Send an email to a contact.
3. Open the email.
4. Check that your contact got X points.
5. Open again the email. See that you got points again.

It is only about **opening emails**
 
### Log errors

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/71239791-problem-feature-with-repeat-point-action?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kuzmany**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | Resolves https://github.com/mautic/mautic/issues/7320
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR try figure out trigger points for each entity in points action reported in this issue https://github.com/mautic/mautic/issues/7320

- added support for events system (eventName instead of callback, callback should be depraced and refactored in future)
- added  new column **internal_id** to point_lead_action_log table - ID or string - in thu future we can store URL for page hit and then we can run points action on each visits
- added new option to email open point action - categories limit and execute one time/each email

image

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create point actions - opens email type
3. Test step by step with different point actions
- Allow repeat the action - should change points everytime
- Execute only: Once - should change points just one time 
- Execute only: For each email - should change points with every email ID - that means If I will create point action just for 3 emails, and all 3 emails my contact open , contact point score adjust point score each time
4. Test also categories filter
 

#### List deprecations along with the new alternative:
1. callback depraced on triggerAction https://github.com/mautic/mautic/blob/0e500aee172676a591341e090873c4120d796a4d/app/bundles/EmailBundle/EventListener/PointSubscriber.php#L82
Use eventName like in other parts of Mautic.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/7330"></a>

### CODE REVIEW

**Comment by johbuch** [comment]
@kuzmany i tried with mautibox but I was not able to get points when opening an email.
Can you check ?
thanks,

**Comment by kuzmany** [comment]
@johbuch try now please

**Review by johbuch — DISMISSED** [review]
I tested all scenarios and everything works now
1) if email point action is repeatable then each time an email is opened you get points. That means if you open the same email 3 times, you are scored 3 times. If you open different emails, you'll get scored as well.
image

2) if email point action is not repeatable, but for each email is checked
image
then every emails opened is scored (but only the first time). That means if you open 3 times the same emails, you will be only scored one time. But if you open other emails (each time the 1st time), you'll be scored
3) if the email point action is executable one time 
image
that means if you open an email, you'll get point, then if you open a second email (different one), you won't get scored.

So now eveything looks good !
thanks zdeno

**Comment by kuzmany** [comment]
Close favor to Mautic 3 branch https://github.com/mautic/mautic/pull/8161

**Comment by kuzmany** [comment]
Reopen for M2 and test this issue https://github.com/mautic/mautic/issues/8603

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@kuzmany looks like we need some test coverage before we can merge this

**Comment by RCheesley** [comment]
@kuzmany do you think you can look at improving the unit test coverage on this PR before the weekend? If not, we'll need to bump it to 3.2.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@kuzmany looks like this is throwing a fatal error causing [tests to fail](https://github.com/mautic/mautic/pull/7330/checks?check_run_id=2422398323#step:9:16). Can you take a look?

File: app/bundles/PointBundle/Event/PointChangeActionExecutedEvent.php
Comment (patrykgruszka) [inline app/bundles/PointBundle/Event/PointChangeActionExecutedEvent.php]: This method doesn't return anything. This PHPDoc should be removed or changed.

File: app/bundles/PointBundle/Event/PointChangeActionExecutedEvent.php
Comment (patrykgruszka) [inline app/bundles/PointBundle/Event/PointChangeActionExecutedEvent.php]: This method doesn't return anything. Return type should be changed to void or removed.

File: app/bundles/EmailBundle/Helper/PointEventHelper.php
Comment (patrykgruszka) [inline app/bundles/EmailBundle/Helper/PointEventHelper.php]: This method always returns bool. The return type should be changed.

**Review by patrykgruszka — CHANGES_REQUESTED** [review]
This is great feature. I've suggested some minor changes and fix for the failing test. I hope you find it useful. I've tested this PR changes and it works as described.

**Comment by OxidoPL** [comment]
This fix would be very helpful when implemented in the upcoming version.

Have a nice day,
Marek

**Comment by patrykgruszka** [comment]
@kuzmany is there any chance of resuming work? It looks like it just needs a minor fix.

**Comment by kuzmany** [comment]
@patrykgruszka We use it on production  for ages. I can fix PHPStan and unit tests, but we need cover new features by unit tests also. At the moment cannot find time for it

**Comment by patrykgruszka** [comment]
@kuzmany I have added unit tests for your PR to cover the new functionality. I have also fixed a bug in the repeatable email opening action, which has also been included in the tests. I've submitted changes as PR to your fork.

**Comment by patrykgruszka** [comment]
@kuzmany I have added unit tests for your PR to cover the new functionality. I have also fixed a bug in the repeatable email opening action, which has also been included in the tests. I've submitted changes as PR to your fork.

**Comment by kuzmany** [comment]
@patrykgruszka thanks. I fixed CS and PHPStan. 
Now looks good to go to 4.3

**Comment by jos0405** [comment]
The PR doesn't give any points.
Tested with different setups, segment email, category, campaign email.

image

**Comment by patrykgruszka** [comment]
@jos0405 are you logged in the Mautic panel in the same browser session? Try to use private session, because points events are not triggered for logged users for some reason.
obraz

**Comment by jos0405** [comment]
Thank you for checking this @patrykgruszka . I figured when the email open "registers", the event is fired.
I will do another test.

**Comment by jos0405** [comment]
I opened this in Gipod, added my own SMTP, sent to my other email address, opened it in ANOTHER browser, where I'm not logged in into gitpod / Mautic.
The opens are registered, but the points still not added. Maybe I'm testing it wrong, plz advice.

**Comment by patrykgruszka** [comment]
@jos0405 what is your point action configuration? 

Here is my simple regression test:
1. Create Segment (S1) with 1 contact
2. Create template email (E1)
3. Create point email action (P1) with settings obraz
4. Create campaign (C1) to send email (E1)
5. Open email in another browser or private session (not logged to mautic panel and without adblock)
6. 3 points should be added to contact 
obraz

I'm using my local DDEV enviroment. There also may be a problem with the GitPod environment. I've tried to test it, but it seems like the mailhog is not running.

**Comment by jos0405** [comment]
I tried all possible combinations, but no luck, however I have the feeling I'm testing it wrong. I'll set up local development env and give it another try. Thx for the tip.

**Review by jos0405 — APPROVED** [review]
I managed to test it in my test environment with real SMTP, not mailhog.
Works really well.
image

I also tested category based points and specific emails.

image

image

image

**Review by jos0405 — APPROVED** [review]
Approved

**Review by jos0405 — DISMISSED** [review]

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php]:
```suggestion
    public function onEmailOpenPointChange(PointChangeActionExecutedEvent $changeActionExecutedEvent): void
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php]:
```suggestion
        if ('email.open' !== $action->getType()) {
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php:139]:
```suggestion
        $triggerMode = $action->getProperties()['triggerMode'] ?? null;
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php]:
```suggestion
        if ('internalId' === $triggerMode) {
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php]:
```suggestion
    public function onEmailSentPointChange(PointChangeActionExecutedEvent $changeActionExecutedEvent): void
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php]:
```suggestion
        if ('email.send' !== $action->getType()) {
```

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php:131]: This comment is not necessary if you'd set the return type for the `changeActionExecutedEvent:: getEventDetails()` method.

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [inline app/bundles/EmailBundle/EventListener/PointSubscriber.php:157]: This comment is not necessary if you'd set the return type for the `changeActionExecutedEvent:: getEventDetails()` method.
```suggestion
```

File: app/bundles/EmailBundle/Form/Type/PointActionEmailOpenType.php
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/PointActionEmailOpenType.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/EmailBundle/Form/Type/PointActionEmailSendType.php
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/PointActionEmailSendType.php:2]:
```suggestion

declare(strict_types=1);

```

**Review by escopecz — COMMENTED** [review]
Thank you for changing the horrible "callback" to events!

I started doing code review and found a lot of small improvements. It makes sense as this is several years old code and you'd write it differently today. I'll stop with the code review and make the small improvements myself as it will save time to both of us. Please shout if you disagree.

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [reply app/bundles/EmailBundle/EventListener/PointSubscriber.php:131]: Looking at the code this makes sense as it can return different entities.

File: app/bundles/EmailBundle/EventListener/PointSubscriber.php
Comment (escopecz) [reply app/bundles/EmailBundle/EventListener/PointSubscriber.php:157]: Looking at the code this makes sense as it can return different entities.

File: app/bundles/PointBundle/Model/PointModel.php
Comment (escopecz) [inline app/bundles/PointBundle/Model/PointModel.php]: @kuzmany question about this. Can we set it for int only? I know you mention this in the description:

> added new column internal_id to point_lead_action_log table - ID or string - in thu future we can store URL for page hit and then we can run points action on each visits

But we cannot store URL to 191 characters. So I think we should go with int for now as that's what we use it for now.

Comment (escopecz) [reply app/bundles/PointBundle/Model/PointModel.php]: I'm also thinking how would this work for existing records in the point_lead_action_log table if the InternalId is not nullable and part of the composed index. Am I missing something?

Comment (escopecz) [reply app/bundles/PointBundle/Model/PointModel.php]: It's also weird that we sometimes send entities like Email or Asset as `$eventDetails` and sometimes entity events like Hit and Submission. This is a mess but cannot change it easily without a BC break.

Comment (escopecz) [reply app/bundles/PointBundle/Model/PointModel.php]: Wait, what is this ID used for again? It's big part of this PR but I don't see any reason for it.

**Review by escopecz — CHANGES_REQUESTED** [review]
I'm going to block this PR for now. I really like that it refactors callbacks to events but I don't get the need for the new internalId column.

I would suggest to remove the InternalId from this PR as I see no reason for it.

@kuzmany please check my commits if you agree with those changes and also check the questions above.

File: app/bundles/PointBundle/Model/PointModel.php
Comment (kuzmany) [reply app/bundles/PointBundle/Model/PointModel.php]: internal_id is set for any ID of item. At the moment it use for store email_id to check If was point trigger on selected emails. 

String we assume support for plugins, If we need check some string ID of external service item.

Comment (escopecz) [reply app/bundles/PointBundle/Model/PointModel.php]: I see. It would be quite confusing to store external ID in a column called internal_id. I think it deserves another column. Or a plugin can create another table to keep whatever data it needs without need for a core table change.

What about the change I did to remove it from the primary key columns and making it nullable. Will it break anything? Why was it primary?

Comment (kuzmany) [reply app/bundles/PointBundle/Model/PointModel.php]: It was build for M2 times ago, then  yes, naming is probably confusing and some code is not the best written.
That column is  something like integration_object_id in sync_object_mapping table, also is varchar. 

Before point action was triggered just for point_id/lead_id one time, this option allow point action trigger in combination point_id/lead_id/object_id. Then you are able to trigger one point action for selected forms, pages etc etc.

Comment (kuzmany) [reply app/bundles/PointBundle/Model/PointModel.php]: Primary key does not make sense on it, of course. Just need index

**Comment by escopecz** [comment]
Ok, we need the field. I can see it now. So what we need here is:
1. Write some functional tests backing up the feature
2. Retest as there were code changes since the last manual tests

**Comment by jos0405** [comment]
It has 1 test already before refactoring. Does this need to be tested again? Or I can set to pending-test confirmation?

**Comment by escopecz** [comment]
@jos0405 yes, it should be tested twice again.

**Comment by patrykgruszka** [comment]
@kuzmany functional tests for this feature are waiting in your repository fork: https://github.com/kuzmany/mautic/pull/36

**Comment by patrykgruszka** [comment]
@escopecz, @kuzmany I think that the primary key for internall_id was necessary. I've got an integrity constraint violation, because of primary key duplicate. That's why the functional test failed.
```
An exception occurred while executing 'INSERT INTO point_lead_action_log (date_fired, internal_id, point_id, lead_id, ip_id) VALUES (?, ?, ?, ?, ?)' with params ["2022-06-27 08:28:46", 2, 1, "1", 2]:

SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY'
```
Any idea how to deal with this? Can we revert this change and make it primary?

There is problem with the PHPSTAN, I've got message:
```
Property Mautic\FormBundle\Tests\Controller\PointSubscriberFunctionalTest::$contactRepository (Mautic\LeadBundle\Entity\LeadRepository) does not accept Doctrine\ORM\EntityRepository<Mautic\LeadBundle\Entity\Lead>.
```
I'm not sure how to deal with this. I've tried `$this->em->getRepository(Lead::class);` and `$this->em->getRepository('MauticLeadBundle:Lead');` which looks okay to me, but fails in PHPSTAN.

File: app/bundles/EmailBundle/Tests/EventListener/PointSubscriberFunctionalTest.php
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/EventListener/PointSubscriberFunctionalTest.php]: Yes, this sucks. I did not find other way around it than something like:
```suggestion
        /** @var LeadRepository $contactRepository;
        $contactRepository = $this->em->getRepository(Lead::class);
        $this->contactRepository  = $contactRepository;
```

File: app/bundles/EmailBundle/Tests/EventListener/PointSubscriberFunctionalTest.php
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/EventListener/PointSubscriberFunctionalTest.php:193]:
```suggestion
    /**
     * @var mixed[] $properties
     */
    protected function createPointAction(array $properties, int $delta, bool $isRepeatable = false): Point
```

**Comment by escopecz** [comment]
I don't get why is it failing. [I removed the primary index](https://github.com/mautic/mautic/pull/7330/commits/6ed2292461b6e4cc9fb714428946f53e287aec8a) from the internal_id field.

**Comment by patrykgruszka** [comment]
@escopecz yes, now the primary key of table `point_lead_action_log` consists of `point_id` and `lead_id`. When you have an action with points for each selected email then you will get a primary key duplicate while saving the second action log just like in test: `PointSubscriberFunctionalTest::testPointsForReadEachSelectedEmail`.
```
-- reads first email
INSERT INTO point_lead_action_log (date_fired, internal_id, point_id, lead_id, ip_id) VALUES ("2022-06-27 08:28:46",1, 1, "1", 2)
-- read second email
INSERT INTO point_lead_action_log (date_fired, internal_id, point_id, lead_id, ip_id) VALUES ("2022-06-27 08:28:47",2, 1, "1", 2)
-- SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY'
```
I think that's why the internal_id was set as part of the primary key before

**Comment by escopecz** [comment]
I see. That makes sense. Can you try to revert the change I made then? We'll have to cover existing records and put some value in via a migration.

**Comment by patrykgruszka** [comment]
@escopecz sure, changes submitted to https://github.com/kuzmany/mautic/pull/37

**Comment by patrykgruszka** [comment]
Ok, tests are looking good now. I've tried to resolve PHPSTAN issues, but I've failed. Maybe you know how to do this right.

File: app/migrations/Version20190319002039.php
Comment (escopecz) [inline app/migrations/Version20190319002039.php:30]: Note to self:

I tested this query on a table with some records. It did not error out but set the value to 0 even though there is no default value defined for this column.

File: app/migrations/Version20190319002039.php
Comment (escopecz) [inline app/migrations/Version20190319002039.php]: Do we need an index when the column is part of the primary key? The primary key is index by itself, right?

Comment (patrykgruszka) [reply app/migrations/Version20190319002039.php]: You are right, we don't need index when column is part of the primary key. Also I see here some weird double drop in this migration. I'll make a change suggestion

Comment (patrykgruszka) [inline app/migrations/Version20190319002039.php]:
```suggestion
        $this->addSql("ALTER TABLE {$tableName} DROP PRIMARY KEY, ADD PRIMARY KEY (`point_id`, `lead_id`, `internal_id`)");
```

File: app/bundles/PointBundle/Entity/LeadPointLog.php
Comment (escopecz) [inline app/bundles/PointBundle/Entity/LeadPointLog.php]: If we all agree that the index is not necessary if the internal_id column is part of the primary key then this should be dropped as well.
```suggestion
            ->setCustomRepositoryClass(LeadPointLogRepository::class);
```

Comment (escopecz) [inline app/bundles/PointBundle/Entity/LeadPointLog.php]: If we all agree that the index is not necessary if the internal_id column is part of the primary key then this should be dropped as well.
```suggestion
            ->setCustomRepositoryClass(LeadPointLogRepository::class);
```

**Comment by patrykgruszka** [comment]
@kuzmany could you please take a look at this PR? I there are only few suggested changes (which are ready to commit) to get this PR ready to test.

**Comment by automatyzuj** [comment]
I will be test this PR. @RCheesley do is chance that this will be merged with 4.4.4? If yes, could you give me a info about deadline.

**Comment by RCheesley** [comment]
For this to be considered for merging it needs to be rebased on 5.x and a duplicate PR made for the 4.4 branch if we want to back-merge it.

Thanks for being willing to help with testing @automatyzuj 🚀

**Comment by patrykgruszka** [comment]
@kuzmany could you please rebase this PR on 5.x? I can help deal with code adjustments if needed.

**Comment by Ilona-Sot** [comment]
@kuzmany  Do you plan to rebase this PR on 5.x? Thank you in advance for the information :)

**Comment by Ilona-Sot** [comment]
@kuzmany If you do not intend to continue working on the task, please let me know, as Comarch we can finish working on this task.

**Comment by kuzmany** [comment]
@Ilona-Sot If you have resources, let's do it. I can assist with finishing it. 
During the fork it for M5 there is chance to rewrite some part of it more simple/logic.

**Comment by RCheesley** [comment]
Hi @Ilona-Sot I wondered if your team might be up for taking this on? It will need more work to implement point groups etc I guess.

**Comment by Ilona-Sot** [comment]
Hi @RCheesley if no one starts the work, we will try to find time for it in Q42024.

**Comment by kuzmany** [comment]
Close at the moment, we need pick up once ready to M5

**Comment by Hugo-Prossaird** [comment]
Update: it's already on a5 -> https://github.com/Webmecanik/Automation_dev/pull/4966

**Pull request closed without merging**
